misk/src/main/kotlin/misk/perf/PauseDetectorThread.kt

```diff
@@ -0,0 +1,155 @@
+package misk.perf
+
+import com.google.common.base.Ticker
+import io.prometheus.client.Summary
+import misk.concurrent.Sleeper
+import misk.metrics.v2.Metrics
+import misk.metrics.v2.PeakGauge
+import org.slf4j.event.Level
+import wisp.logging.getLogger
+import java.time.Duration
+import java.util.concurrent.TimeUnit.MILLISECONDS
+import java.util.concurrent.TimeUnit.NANOSECONDS
+import javax.inject.Inject
+import javax.inject.Singleton
+
+/**
+ * Detects and records pauses experienced by the VM. Garbage collection is a common source of
+ * pauses, but pauses can be caused by many things: CPU throttling, excessive swapping, excessive
+ * active threads, etc.
+ *
+ * Heavily inspired by [jhiccup](https://www.azulsystems.com/jHiccup) which is
+ * [public domain](https://creativecommons.org/publicdomain/zero/1.0/)
+ */
+@Singleton
+class PauseDetectorThread @Inject constructor(
+  private val config: PauseDetectorConfig,
+  private val ticker: Ticker,
+  private val sleeper: Sleeper,
+  val metrics: Metrics,
+) : Thread() {
+  private val logger = getLogger<PauseDetectorThread>()
+
+  /** Log levels by pause time sorted by severity descending*/
+  private val logLevels: List<LogLevels>
+
+  /**
+   * Tracks the distribution and total amount of pause time observed.
+   *
+   * (Using a summary instead of a histogram to prefer more accurate node-local data over
+   * more accurate aggregations)
+   */
+  private val pauseSummary: Summary
+
+  /** Tracks peak pause time. */
+  private val pausePeak: PeakGauge
+
+  /* If false, do not schedule another iteration. */
+  @Volatile
+  private var keepRunning = true
+
+
+  // No synchronization is necessary for these variables: they are only ever accessed by the
+  // detector thread itself OR by a test harness thread.
+  //
+  // (They are also initialized in the constructor, but that initialization has a happens-before
+  // relationship to the [Thread.start] that starts the thread).
+  /** Used to prevent generating early noise (for example in the first few seconds of startup) */
+  private var shortestObservedDeltaTimeNsec = Long.MAX_VALUE
+  /** Tracks the start time of the last invocation to [sleep] */
+  private var startTime: Long = 0
+
+  init {
+    isDaemon = true
+    name = "PauseDetector"
+
+    logLevels = listOf(
+      LogLevels(config.logErrorMillis, Level.ERROR),
+      LogLevels(config.logWarnMillis, Level.WARN),
+      LogLevels(config.logInfoMillis, Level.INFO)
+    )
+
+    pauseSummary =
+      metrics.summary("jvm_pause_time_summary_ms", "Summary in ms of pause time", listOf())
```

Comment:
Seems like this fits with our metric conventions, but please let me know if there's a better name structure to use here!

FWIW `framework.perf.pause` is the metric path used by the Square version